### PR TITLE
refactor structure

### PR DESCRIPTION
**Pull request by Moelf**

### CODE REVIEW

File: src/PhotosplineReader.jl [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            # local_basis[:, d] .= bsplinebasisall( BSplineDerivativeSpace{1}(bsp_space), centers[d]-1, x[d] )
+
+        else
+
+            # handle edges... see https://github.com/icecube/photospline/blob/master/include/photospline/bspline.h#L44
+            if ( centers[d]-1 == orders[d] ) && ( x[d] < knots[d][centers[d]] ) 
+                local_basis[ 1, d ] = bsplinebasis( bsp_space, 1, x[d] )
+                local_basis[ 2:(orders[d]+1), d ] .= 0.
+
+            elseif ( centers[d] == x[d] - ( orders[d] + 1) ) && ( x[d] > knots[d][centers[d]+1] )
```
Comment (Moelf) [inline src/PhotosplineReader.jl:158]: this `x[d]` here was `spt.nknots[d]`, you should double-check if this is the correct variable

Comment (kcarloni) [reply src/PhotosplineReader.jl:158]: I think `spt.nknots[d]` is correct, although the boundary condition handling is kind of tricky. basically it's checking whether the relevant subset of the knots (`centers[d]`) is close to the right edge of the knots. see https://github.com/icecube/photospline/blob/5b4ff896780f810768189bd717d02a0a033b8cb8/include/photospline/bspline.h#L51

Comment (Moelf) [reply src/PhotosplineReader.jl:158]: this function's signature is:
```julia
function ndsplineeval( x, centers, derivatives_bool, knots, coeffs, orders, nd )
```

so `spt` originally was giving me variable undefined error

Comment (kcarloni) [reply src/PhotosplineReader.jl:158]: oh, thanks! good catch.

**Pull request merged**